src/core/cgroup.c

```diff
@@ -3989,6 +3989,38 @@ int unit_get_ip_accounting(
         return r;
 }
 
+static uint64_t unit_get_eff_limit_one(Unit *u, CGroupLimitType type) {
+        CGroupContext *cc;
+
+        assert(u);
+        assert(UNIT_HAS_CGROUP_CONTEXT(u));
+
+        static const size_t members[_CGROUP_LIMIT_MAX] = {
+                [CGROUP_LIMIT_MEMORY_MAX]    = offsetof(CGroupContext, memory_max),
+                [CGROUP_LIMIT_MEMORY_HIGH]   = offsetof(CGroupContext, memory_high),
+                [CGROUP_LIMIT_TASKS_MAX]     = offsetof(CGroupContext, tasks_max),
+        };
+
+        cc = unit_get_cgroup_context(u);
+        return *(uint64_t *)((uint8_t *)cc + members[type]);
+}
+
+int unit_get_eff_limit(Unit *u, CGroupLimitType type, uint64_t *ret) {
+        uint64_t min;
+        assert(u);
+
+        if (!UNIT_HAS_CGROUP_CONTEXT(u))
+                return -EINVAL;
+
```

Comment:
hmm, shouldn#t this ultimately check the physical amount of memory available as final cap? i.e. something like physical_memory() (though maybe incorporating swap too).

similar for system_tasks_max() for the effective tasks limit

Reply:
I notice that `MemoryAvailable=` (that @bluca pointed out) doesn't take global limit into account. I think it'd be good to have the two calculations consistent. 
Actually, if anti-consistent, the other way of crossing would make more sense to me:
- available -- that should take the global resource _availability_ into account
- effective limit -- that should take into account _limits_ on the path

(But I would rather not advertise `MemoryAvailable=` that much because users should rather watch memory PSI for (self-)control.)

With that said, I'm fine with either variant of what incorporate into the effective limit, current version includes the global.

Reply:
I think we should keep them consistent, and not check physical memory. After all, this is about cgroup resources and accounting, so we should stick to that.

Reply:
but i think the name "effective" is misleading then. if you don#t take the physical limits into account, then the resulting values are not effective, but theoretic...

Hence, maybe call this something else, such as "Combined"? or "Subsequent", "Overall"? "Resulting"?

Reply:
i'd also be fine with tweaking MemoryAvailable= btw

Reply:
so, I think i#d be fine with leaving the physical RAM out of the equation, but we definitely should take the root cgroup settings (which matters only for containers I guess) into account. (but the patch already does?)

but the name should prbably be changed then, because I don't think this is the "effective" limit then, but just the combined limit or so.